src/storage/src/source/mod.rs

```diff
@@ -974,94 +952,101 @@ where
-            .iter()
-            .cloned()
-            .flat_map(|(pid, offset)| Some((pid, offset?)))
-            .collect();
-
-        let mut source_reader = if !active {
-            None
-        } else {
-            match S::new(
-                name.clone(),
+
+        let sync_activator = scope.sync_activator_for(&info.address[..]);
+        let base_metrics = base_metrics.clone();
+        let source_connector = source_connector.clone();
+        let mut source_reader = Box::pin(async_stream::stream! {
+            let mut timestamper = reclock::Timestamper::new(storage_metadata, as_of, now).await;
+
+            // TODO(petrosagg): What is the purpose of an partition offset being None?
+            let start_offsets = timestamper.partition_cursors().into_iter().map(|(p, o)| (p, Some(o))).collect();
+
+            let source_reader = S::new(
+                name,
                 id,
                 worker_id,
                 worker_count,
-                scope.sync_activator_for(&info.address[..]),
+                sync_activator,
                 source_connector.clone(),
                 aws_external_id.clone(),
-                restored_offsets,
+                start_offsets,
                 encoding,
-                base_metrics.clone(),
-            ) {
-                Ok(source_reader) => Some(source_reader.into_stream(timestamp_frequency)),
+                base_metrics,
+            );
+            let source_stream = match source_reader {
+                Ok(s) => s.into_stream(timestamp_frequency).fuse(),
                 Err(e) => {
                     error!("Failed to create source: {}", e);
-                    None
-                }
-            }
-        };
-
-        move |cap, durability_cap, output| {
-            // First check that the source was successfully created
-            let source_reader = match &mut source_reader {
-                Some(source_reader) => source_reader,
-                None => {
-                    return SourceStatus::Done;
+                    return;
                 }
             };
 
-            // Check that we have a valid list of timestamp bindings.
-            let timestamp_histories = match &mut timestamp_histories {
-                Some(histories) => histories,
-                None => {
-                    error!("Source missing list of timestamp bindings");
-                    return SourceStatus::Done;
+            tokio::pin!(source_stream);
+
+            let mut timestamp_interval = tokio::time::interval(timestamp_frequency);
+            let mut pending_messages = vec![];
+            loop {
+                tokio::select! {
+                    item = source_stream.next() => {
+                        match item {
+                            Some(Ok(message)) => pending_messages.push(message),
+                            Some(Err(_)) => {
+                                panic!("Errors have no offset, and so cannot be remapped");
```

Comment:
wait, aren't these errors normal errors that sources can output?

Reply:
Yeah this is a definiteness bug i realized while working on this PR. What timestamp should an error with no offset be assigned to?

Reply:
@petrosagg assign it the offset of the most recent message? (or 0 if there is none)?

Reply:
But are we guaranteed to see the same error if we resume reading the source from this offset? I feel like we need to change SourcerReader and make errors have offsets set by the source implementation, and require that they must be definite, otherwise we can get into trouble

Reply:
hmm thats a good point. what do we do now with the error stream from sources? do we not persist timestamps for them, and instead just error the source and give up?

Reply:
We just assign the current timestamp and hope for the best